lib/virtual-fs/src/mem_fs.rs

```diff
@@ -0,0 +1,600 @@
+    }
+}
+impl Seek for MemFileHandle {
+    fn seek(&mut self, pos: io::SeekFrom) -> io::Result<u64> {
+        let mut inner = self.fs.inner.lock().unwrap();
+        let file = inner
+            .inodes
+            .get_mut(&self.inode)
+            .ok_or_else(Self::no_file_err)?;
+
+        file.seek(pos)
+    }
+}
+impl Write for MemFileHandle {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut inner = self.fs.inner.lock().unwrap();
+        let file = inner
+            .inodes
+            .get_mut(&self.inode)
+            .ok_or_else(Self::no_file_err)?;
+
+        file.write(buf)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        let mut inner = self.fs.inner.lock().unwrap();
+        let file = inner
+            .inodes
+            .get_mut(&self.inode)
+            .ok_or_else(Self::no_file_err)?;
+
+        file.flush()
+    }
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+        let mut inner = self.fs.inner.lock().unwrap();
+        let file = inner
+            .inodes
+            .get_mut(&self.inode)
+            .ok_or_else(Self::no_file_err)?;
+
+        file.write_all(buf)
+    }
+    fn write_fmt(&mut self, fmt: ::std::fmt::Arguments) -> io::Result<()> {
+        let mut inner = self.fs.inner.lock().unwrap();
+        let file = inner
+            .inodes
+            .get_mut(&self.inode)
+            .ok_or_else(Self::no_file_err)?;
+
+        file.write_fmt(fmt)
+    }
+}
+
+#[typetag::serde]
+impl VirtualFile for MemFileHandle {
+    fn last_accessed(&self) -> u64 {
+        let inner = self.fs.inner.lock().unwrap();
+        inner
+            .inodes
+            .get(&self.inode)
+            .as_ref()
+            .map(|file| file.last_accessed())
+            .unwrap_or_default()
+    }
+
+    fn set_last_accessed(&self, last_accessed: u64) {
+        let inner = self.fs.inner.lock().unwrap();
+        inner
+            .inodes
+            .get(&self.inode)
+            .as_ref()
+            .map(|file| file.set_last_accessed(last_accessed));
+    }
+
+    fn last_modified(&self) -> u64 {
+        let inner = self.fs.inner.lock().unwrap();
+        inner
+            .inodes
+            .get(&self.inode)
+            .as_ref()
+            .map(|file| file.last_modified())
+            .unwrap_or_default()
+    }
+
+    fn set_last_modified(&self, last_modified: u64) {
+        let inner = self.fs.inner.lock().unwrap();
+        inner
+            .inodes
+            .get(&self.inode)
+            .as_ref()
+            .map(|file| file.set_last_modified(last_modified));
+    }
+
+    fn created_time(&self) -> u64 {
+        let inner = self.fs.inner.lock().unwrap();
+        inner
+            .inodes
+            .get(&self.inode)
+            .as_ref()
+            .map(|file| file.created_time())
+            .unwrap_or_default()
+    }
+
+    fn set_created_time(&self, created_time: u64) {
+        let inner = self.fs.inner.lock().unwrap();
```

Comment:
If `set_created_time` uses `&mut self` we can move from interior mutability to exterior very easily (so no locks are needed)

Reply:
In this case we need still need it because it's shared. Interior mutability wouldn't work unless we use atomics. Which _is_ possible here but probably not worth it

Reply:
Yeah if we make `OpenOptions` have a lifetime then maybe that can work  if we change the whole api to use `&mut` and require locking from API users (like our existing wasmer code)